urlac47192b66_ggrandes_bouncer_tgz-pJ8-org_javastack_bouncer_ClusterTestJ8/src/org/javastack/bouncer/InboundAddress.java

```diff
@@ -0,0 +1,80 @@
+package org.javastack.bouncer;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.UnknownHostException;
+
+/**
+ * Representation of listen address
+ */
+class InboundAddress extends BouncerAddress {
+	final String host;
+	final int port;
+	final Options opts;
+
+	SSLFactory sslFactory = null;
+	InetAddress[] addrs = null;
+
+	InboundAddress(final ServerContext context, final String host, final int port, final Options opts) {
+		super(context);
+		this.host = host;
+		this.port = port;
+		this.opts = opts;
+	}
+
+	@Override
+	void setSSLFactory(final SSLFactory sslFactory) {
+		this.sslFactory = sslFactory;
+	}
+
+	@Override
+	Options getOpts() {
+		return opts;
+	}
+
+	@Override
+	public String toString() {
+		return (host.equals("0.0.0.0") ? "*" : host) + ":" + port;
+	}
+
+	@Override
+	void resolve() throws UnknownHostException {
+		if (checkUpdateResolv()) {
+			addrs = InetAddress.getAllByName(host);
+			Log.info(this.getClass().getSimpleName() + " Resolved " + String.valueOf(this) + //
+					" [" + fromArrAddress(addrs) + "]");
+		} else {
+			Log.info(this.getClass().getSimpleName() + " Resolve (cached) " + String.valueOf(this) + //
+					" [" + fromArrAddress(addrs) + "]");
+		}
+	}
+
+	InetSocketAddress[] getSocketAddress() {
+		final InetSocketAddress[] socks = new InetSocketAddress[addrs.length];
+		for (int i = 0; i < socks.length; i++) {
+			socks[i] = new InetSocketAddress(addrs[i], port);
+		}
+		return socks;
+	}
+
+	ServerSocket listen() throws IOException {
+		resolve();
+		final ServerSocket listen;
+		if (opts.isOption(Options.MUX_SSL | Options.TUN_ENDSSL)) {
+			listen = sslFactory.createSSLServerSocket();
+		} else {
+			listen = new ServerSocket();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource: *listen*. The resource is referenced at lines 71, 73, 74. The resource is returned at line 78. There are other execution paths that don&#39;t close the resource or return it, for example, when ServerSocket.bind throws an exception. To prevent this resource leak, close *listen* along these other paths before you exit this method.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).